File: 프로그래머스/Level3/추석 트래픽/solution.py
def solution(lines):
    logs = []
    for line in lines:
        l = len(line)
        end = int(line[11:13]) * 3600 * 1000 + int(line[14:16]) * 60 * 1000 + int(line[17:19]) * 1000 + int(line[20:23])
        start = end - int(line[24]) * 1000 + 1
        if l > 27:
            start -= int(line[26]) * 100
        if l > 28:
            start -= int(line[27]) * 10
        if l > 29:
            start -= int(line[28])
        
        logs.append((start, end))
    
    ans = 0
    for i in range(len(logs)):
        start = 0
        end = 0
        for j in range(len(logs)):
            if logs[i][0] <= logs[j][0] < logs[i][0] + 1000:
                start += 1
            elif logs[i][0] <= logs[j][1] < logs[i][0] + 1000:
                start += 1
            elif logs[j][0] <= logs[i][0] <= logs[j][1]:
                start += 1
            
            if logs[i][1] <= logs[j][0] < logs[i][1] + 1000:
                end += 1
            elif logs[i][1] <= logs[j][1] < logs[i][1] + 1000:
                end += 1
            elif logs[j][0] <= logs[i][1] <= logs[j][1]:
                end += 1
        ans = max(ans, max(start, end))
        
    return ans

File: 프로그래머스/Level3/추석 트래픽/solution.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
#define ll long long
#define pli pair<ll, int>

using namespace std;

vector<ll> endTime, process;

void divide(string line)
{
    ll start = ((line[11] - '0') * 10 + (line[12] - '0')) * 1000 * 60 * 60;
    start += ((line[14] - '0') * 10 + (line[15] - '0')) * 1000 * 60;
    start += ((line[17] - '0') * 10 + (line[18] - '0')) * 1000;
    start += ((line[20] - '0') * 100 + (line[21] - '0') * 10 + (line[22] - '0'));

    endTime.push_back(start);

    process.push_back(0);
    process.back() += (line[24] - '0') * 1000;
    for (int i = 26, val = 100; line[i] != 's' && line[i] != '\0'; i++, val /= 10)
    {
        process.back() += (line[i] - '0') * val;
    }
}

int isRange(const ll &sTime, const ll &csTime, const ll &ceTime)
{
    if (sTime <= csTime && csTime <= sTime + 999)
        return 1;
    if (sTime <= ceTime && ceTime <= sTime + 999)
        return 1;
    if (csTime <= sTime && sTime + 999 <= ceTime)
        return 1;
    return 0;
}

int solution(vector<string> lines)
{
    for (int i = 0; i < lines.size(); i++)
    {
        divide(lines[i]);
    }

    int ans = 0;

    for (int i = 0, sCnt = 0, eCnt = 0; i < lines.size(); i++, sCnt = 0, eCnt = 0)
    {
        ll sTime = endTime[i] - process[i] + 1;
        for (int j = 0; j < lines.size(); j++)
        {
            ll csTime = endTime[j] - process[j] + 1;
            sCnt += isRange(sTime, csTime, endTime[j]);
            eCnt += isRange(endTime[i], csTime, endTime[j]);
        }
        ans = max(ans, max(sCnt, eCnt));
    }

    return ans;
}